### PR TITLE
mcobbett refactor events publishing

### PR DESCRIPTION
**Pull request by techcobweb**
# Why ?

Event producing/publishing within the framework component can be done using a dedicated
interface.

This is part of the general refactoring of the test runners to remove some bad code smells.

In this case, it's having duplicated function, and no class which gathers together all of the usage of the IEventProducer interface.

- **Refactored events publishing to collect events via an interface**

### CODE REVIEW

**Comment by galasa-team** [comment]
Build successful

File: galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/events/IEventsPublisher.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ */
+public interface IEventsPublisher {
+    void setEnabled(boolean isProduceEventsEnabled);
+
+    boolean isEnabled();
+
+    void publishTestHeartbeatStoppedEvent(String testRunName) throws TestRunException ;
+
+    void publishTestRunLifecycleStatusChangedEvent(String testRunName, TestRunLifecycleStatus status) throws TestRunException ;
+}
```
[the remark this answers is not in the document]
Comment (techcobweb) [reply galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/events/IEventsPublisher.java:22]: I did see the IEvent producer interface, which is general-purpose in terms of topic and payload. I wanted something more specific to the events that framework can publish sat on top of it.

Perhaps if I renamed IEventsPublisher to IFrameworkEventsProducer that would be better ?

File: galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/events/FrameworkEventsServiceRegistration.java [hunk 2]
```diff
@@ -31,6 +31,7 @@ public void initialise(@NotNull IFrameworkInitialisation frameworkInitialisation
             URI cps = frameworkInitialisation.getBootstrapConfigurationPropertyStore();
 
             // If the CPS is a file, then register this version of the EventsService
+            // TODO: Why ? Don't we want to create events when the CPS is on the cluster ? We should explain why if this is the case.
```
[the remark this answers is not in the document]
Comment (techcobweb) [reply galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/events/FrameworkEventsServiceRegistration.java:34]: If extensions or a manager wanted to send events, they would need the IEventsProducer service to do so, but not the framework-specific things.

[the remark this answers is not in the document]
Comment (techcobweb) [reply galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/events/FrameworkEventsServiceRegistration.java:34]: Oh. The source of my confusion is that I thought it was registering the real implementation only if there was a file. Maybe I should change the name of the framework implementation from `FrameworkEventsService` to `NullEventsService` or something to indicate what it does ?

File: galasa-parent/dev.galasa.framework/src/test/java/dev/galasa/framework/mocks/MockEventsPublisher.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package dev.galasa.framework.mocks;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import dev.galasa.framework.TestRunException;
+import dev.galasa.framework.TestRunLifecycleStatus;
+import dev.galasa.framework.internal.events.IEventsPublisher;
+
+public class MockEventsPublisher implements IEventsPublisher {
```
[the remark this answers is not in the document]
Comment (techcobweb) [reply galasa-parent/dev.galasa.framework/src/test/java/dev/galasa/framework/mocks/MockEventsPublisher.java:10]: I needed to record the events being produced, and assert that they were coming out and in order...etc. So needed an implementation.

Lets discuss the need or not for this framework-specific interface on top of the generic-interface.

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

File: galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/events/EventsPublisher.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+ * An implementation of something which publishes events through the OSGi registered events service.
+ */
+public class EventsPublisher implements IEventsPublisher {
+
+    private Log logger = LogFactory.getLog(EventsPublisher.class);
+    
+    private IConfigurationPropertyStoreService cps;
+    private IEventsService eventsService ;
+    private boolean isEnabled = false;
+
```
[the remark this answers is not in the document]
Comment (techcobweb) [reply galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/events/EventsPublisher.java]: Discussed about needing something framework-specific to keep even logic together above the IEvent interface.

**Review by jadecarino — APPROVED** [review]

**Comment by galasa-team** [comment]
Build successful

**Pull request merged**